pypeit/flatfield.py

```diff
@@ -110,21 +130,59 @@ def _bundle(self):
             if self[key] is None:
                 continue
             # Array?
-            if self.datamodel[key]['otype'] == np.ndarray and key != 'spat_bsplines':
+            if self.datamodel[key]['otype'] == np.ndarray and \
+                    key != 'pixelflat_spat_bsplines' and \
+                    key != 'illumflat_spat_bsplines':
                 d.append({key: self[key]})
-            elif key == 'spat_bsplines':
+            elif key == 'pixelflat_spat_bsplines':
                 for ss, spat_bspl in enumerate(self[key]):
                     # Naming
-                    spat_bspl.hdu_prefix = 'SPAT_ID-{}_'.format(self.spat_id[ss])
+                    spat_bspl.hdu_prefix = 'PIXELFLAT_SPAT_ID-{}_'.format(self.spat_id[ss])
                     dkey = 'bspline-{}'.format(self.spat_id[ss])
                     # Save
-                    d.append({dkey: spat_bspl})
+                    d.append(copy.deepcopy({dkey: spat_bspl}))
+            elif key == 'illumflat_spat_bsplines':
+                for ss, spat_bspl in enumerate(self[key]):
+                    # Naming
+                    spat_bspl.hdu_prefix = 'ILLUMFLAT_SPAT_ID-{}_'.format(self.spat_id[ss])
+                    dkey = 'bspline-{}'.format(self.spat_id[ss])
+                    # Save
+                    d.append(copy.deepcopy({dkey: spat_bspl}))
             else: # Add to header of the primary image
                 d[0][key] = self[key]
         # Return
         return d
 
-    def fit2illumflat(self, slits, initial=False, flexure_shift=None):
+    @property
+    def procflat(self):
+        return self.pixelflat_raw
+
+    @property
+    def pixelflat(self):
+        return self.pixelflat_norm
+
+    @property
+    def flat_model(self):
+        return self.pixelflat_model
+
+    @property
+    def bpmflats(self):
+        if self.pixelflat_bpm is not None:
+            return self.pixelflat_bpm
+        elif self.illumflat_bpm is not None:
+            return self.illumflat_bpm
+        else:
+            msgs.warn("FlatImages contains no BPM - trying to generate one")
+            return np.zeros(self.procflat.shape, dtype=np.int)
+
+    @property
+    def spat_bsplines(self):
+        if self.illumflat_spat_bsplines is not None:
+            return self.illumflat_spat_bsplines
+        else:
+            return self.pixelflat_spat_bsplines
+
+    def fit2illumflat(self, slits, frametype='illum', initial=False, flexure_shift=None):
```

Comment:
This frametype argument that you are adding is not documented. I cannot see any reason for it. Can you please explain why this code needs to know the frametype? I thought we agreed you would do illum/pixel specific things outside of this class?